echopype/clean/utils.py

```diff
@@ -0,0 +1,228 @@
+        # any values prior to the first bin.
+        np.digitize(ds_Sv["depth"], left_bin_values) - 1,
+        dims=["channel", "ping_time", "range_sample"],
+    )
+
+    # Upsample the downsampled Sv
+    upsampled_Sv = xr.apply_ufunc(
+        _upsample_using_mapping,
+        # Need to compute arrays since indexing fails when delayed:
+        downsampled_Sv.compute(),
+        upsample_mapping.compute(),
+        input_core_dims=[["depth_bins"], ["range_sample"]],
+        output_core_dims=[["range_sample"]],
+        exclude_dims=set(["depth_bins", "range_sample"]),
+        dask="parallelized",
+        vectorize=True,
+        output_dtypes=[np.float64],
+    )
+    # Set range sample coords
+    upsampled_Sv = upsampled_Sv.assign_coords(range_sample=ds_Sv["range_sample"])
+
+    return downsampled_Sv, upsampled_Sv
+
+
+def echopy_impulse_noise_mask(
+    Sv: np.ndarray, num_side_pings: int, impulse_noise_threshold: float
+) -> np.ndarray:
+    """Single-channel impulse noise mask computation from echopy."""
+    # Construct the two ping side-by-side comparison arrays
+    dummy = np.zeros((Sv.shape[0], num_side_pings)) * np.nan
+    comparison_forward = Sv - np.c_[Sv[:, num_side_pings:], dummy]
+    comparison_backward = Sv - np.c_[dummy, Sv[:, 0:-num_side_pings]]
+    comparison_forward[np.isnan(comparison_forward)] = np.inf
+    comparison_backward[np.isnan(comparison_backward)] = np.inf
+
+    # Create mask by checking if comparison arrays are above `impulse_noise_threshold`
+    maskf = comparison_forward > impulse_noise_threshold
+    maskb = comparison_backward > impulse_noise_threshold
+    mask = maskf & maskb
+
+    return mask
+
+
+def echopy_attenuated_signal_mask(
+    Sv: np.ndarray,
+    depth: np.ndarray,
+    upper_limit_sl: float,
+    lower_limit_sl: float,
+    num_pings: int,
+    attenuation_signal_threshold: float,
+) -> np.ndarray:
+    """Single-channel attenuated signal mask computation from echopy."""
+    # Initialize mask
+    attenuated_mask = np.zeros(Sv.shape, dtype=bool)
+
+    for ping_time_idx in range(Sv.shape[0]):
+
+        # Find indices for upper and lower SL limits
+        up = np.argmin(abs(depth[ping_time_idx, :] - upper_limit_sl))
+        lw = np.argmin(abs(depth[ping_time_idx, :] - lower_limit_sl))
+
+        # Mask when attenuation masking is feasible
+        if not (
+            (ping_time_idx - num_pings < 0)
+            | (ping_time_idx + num_pings > Sv.shape[0] - 1)
+            | np.all(np.isnan(Sv[ping_time_idx, up:lw]))
+        ):
+            # Compare ping and block medians, and mask ping if difference greater than
+            # threshold.
+            pingmedian = _lin2log(np.nanmedian(_log2lin(Sv[ping_time_idx, up:lw])))
+            blockmedian = _lin2log(
+                np.nanmedian(
+                    _log2lin(Sv[(ping_time_idx - num_pings) : (ping_time_idx + num_pings), up:lw])
```

Comment:
so here `num_pings` actually corresponds to $n/2$ in the $S_{v_n}$ in equation 2 -- is this "inherited" from echopy?

Reply:
Yeah, this is inherited directly from `echopy`. Maybe instead of `num_ping` I'll rename this to `num_side_pings`, following the convention from transient noise masking. I think it's kind of hard to work with this `n/2` since we would have to enforce even numbers and the user would have to do that `n/2` calculation in their head.